### PR TITLE
Initial implementation of toggle button

### PR DESCRIPTION
**Pull request by ksin**
Initial implementation for https://github.com/Ticketfly-UI/issues/issues/5

Feel free to clone down and serve the dummy to play with. Will put the dummy app in the github-pages for this once this PR is merged.

@BrianSipple @melissaroman @spencer516 

screen shot 2017-02-10 at 5 01 41 pm
screen shot 2017-02-10 at 5 01 47 pm

### CODE REVIEW

File: addon/components/tf-toggle-button.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+  tagName: 'button',
+  type: 'button',
+  attributeBindings: ['disabled', 'aria-label', 'aria-pressed', 'type'],
+  classNameBindings: ['buttonCategory', 'disabled:c-tf-toggle-button--disabled', 'pressed:c-tf-toggle-button--pressed'],
+  classNames: ['c-tf-toggle-button', 'u-pv-0', 'u-ph-xs'],
+
+  hook: 'tf-toggle-button',
+
+  onClick() { return this; }, //noop
```
Comment (spencer516) [inline addon/components/tf-toggle-button.js]: No need to `return this;`

File: addon/components/tf-toggle-button.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+  click() {
+    this.toggleProperty('pressed');
```
Comment (spencer516) [inline addon/components/tf-toggle-button.js]: Is `pressed` the right name for the boolean here? Maybe `toggled`? Or, parity with checkbox: `checked`?

File: addon/components/tf-toggle-button.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    const pressed = get(this, 'pressed');
+    get(this, 'onClick')(pressed);
```
Comment (spencer516) [inline addon/components/tf-toggle-button.js]: I think a good convention we could adopt is 
```js
Ember.tryInvoke(this, 'onClick', pressed)
```

Comment (spencer516) [reply addon/components/tf-toggle-button.js]: Also — is `onClick` the right action name for this? That implies that the argument that gets passed will be the click event...perhaps we should just support `action`?

File: addon/components/tf-toggle-button.js (hunk 2 above)
Comment (spencer516) [reply addon/components/tf-toggle-button.js]: (Ahh — I see that this corresponds with `aria-pressed`... 🤔 )

File: addon/components/tf-toggle-button.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  },
+
+  pressed: false,
+
+  // aria-pressed is not a boolean value, but a string token of 'true' or 'false'
+  "aria-pressed": computed('pressed', {
```
Comment (spencer516) [inline addon/components/tf-toggle-button.js]: I might be totally wrong, but _I think_ we can camelize this attribute to be `ariaPressed` and it "just works". If I am wrong, single quotes here?

File: app/styles/components/tf-toggle-button.css [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+*/
+
+.c-tf-toggle-button {
+  border: var(--tf-tb-border);
+  border-radius: var(--tf-tb-border-radius);
+  -moz-box-shadow:    var(--tf-tb-box-shadow);
+  -webkit-box-shadow: var(--tf-tb-box-shadow);
+  box-shadow:         var(--tf-tb-box-shadow);
+  background-color: var(--x-color-gray-1);
+  text-align: left;
```
Comment (spencer516) [inline app/styles/components/tf-toggle-button.css:33]: I'm going to guess that we're going to want to animate that button at some point — and, I don't think flipping `text-align` from `left` to `right` is going to be animatable. There's a few ways we could do this — let's discuss.

**Review by spencer516 — COMMENTED** [review]
Thanks for moving this forward❗️❗️ 

Added some thoughts ^^

File: addon/components/tf-toggle-button.js (hunk 2 above)
Comment (ksin) [reply addon/components/tf-toggle-button.js]: I'm open to `toggled` or `pressed`

File: app/styles/components/tf-toggle-button.css (hunk 5 above)
Comment (ksin) [reply app/styles/components/tf-toggle-button.css:33]: Agreed and aware of this. Just wanted to get something out easiest as possible until I get some input from you / designers.

File: addon/components/tf-toggle-button.js (hunk 4 above)
Comment (ksin) [reply addon/components/tf-toggle-button.js]: only true for classNameBindings

**Comment by ksin** [comment]
@spencer516 updated `pressed`. Pay special attention to 
https://github.com/Ticketfly-UI/ember-ticketfly-toggle-button/pull/1/commits/a1c6aed7a4fda9c15643eac10561344ba3e23f88#diff-26fc9f31eee515f75cc90b2c7e31f4b2R56

https://github.com/Ticketfly-UI/ember-ticketfly-toggle-button/pull/1/commits/a1c6aed7a4fda9c15643eac10561344ba3e23f88#diff-794fdeab8227bca2614dbb65b5cb56c7R97

https://github.com/Ticketfly-UI/ember-ticketfly-toggle-button/pull/1/commits/a1c6aed7a4fda9c15643eac10561344ba3e23f88#diff-794fdeab8227bca2614dbb65b5cb56c7R85

File: addon/components/tf-toggle-button.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+  colorGroup: '',
+  sizeGroup: 'primary',
+
+  buttonCategory: computed('colorGroup','sizeGroup', {
+    get() {
+      const sizeGroup = get(this, 'sizeGroup');
+
+      let buttonCategory = `c-tf-toggle-button--${sizeGroup}`;
+
+      const colorGroup = get(this, 'colorGroup');
```
Comment (spencer516) [inline addon/components/tf-toggle-button.js]: I think we should probably make these more single-purpose computed properties — so, `buttonCategory` is one and `colorCategory` is another.

File: addon/components/tf-toggle-button.js [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+  }),
+
+  /*
+    `_pressed` is an internal state of the toggle button. `_pressed` will be set to the passed in `pressed` param. 
+    Clicking the component will update `_pressed`, but it won't mutate `pressed`. It will send an action with the 
+    updated value of `_pressed`.
+
+    This is to ensure that the toggle button doesn't mutate the passed in value of `pressed`. Only
+    the consumer can do that, in accordance with DDAU. This also allows the component to function on its own, 
+    without requiring the consumer to define and update its pressed state every time it is used.
```
Comment (spencer516) [inline addon/components/tf-toggle-button.js:62]: ^ 👍 Good and helpful comment!

File: tests/integration/components/tf-toggle-button-test.js [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+test(`mut helper to update the value of 'pressed'`, function(assert) {
+  this.set('pressed', true);
+
+  this.render(hbs`{{tf-toggle-button pressed=pressed action=(action (mut pressed))}}`);
+
+  $hook('tf-toggle-button').click();
+
+  const pressed = this.get('pressed');
+
+  assert.notOk(pressed, 'Pressed is false, mutated by the mut helper action');
```
Comment (spencer516) [inline tests/integration/components/tf-toggle-button-test.js:106]: 👍

**Review by spencer516 — COMMENTED** [review]
I think just one other note — but this is looking good.

**Pull request merged**